### PR TITLE
Add support for message box

### PR DESCRIPTION
**Pull request by uyjulian**
Allow `SDL_ShowMessageBox` and `SDL_ShowSimpleMessageBox` to be used.

### CODE REVIEW

File: src/render/vitagxm/SDL_render_vita_gxm_tools.c [hunk 1]
```diff
@@ -1178,6 +1178,77 @@ gxm_texture_set_filters(gxm_texture *texture, SceGxmTextureFilter min_filter, Sc
...
+            VITA_GXM_SCREEN_STRIDE,
+            buffer_for_common_dialog[i].displayData.address
+        );
+        sceGxmSyncObjectCreate(&buffer_for_common_dialog[i].sync);
+    }
+}
+
+void gxm_swap_for_common_dialog(void)
+{
+    sceGxmPadHeartbeat(&buffer_for_common_dialog[back_buffer_index_for_common_dialog].surf, buffer_for_common_dialog[back_buffer_index_for_common_dialog].sync);
```
Comment (isage) [inline src/render/vitagxm/SDL_render_vita_gxm_tools.c]: Do you really need that?

**Comment by isage** [comment]
That looks good. Is that still WIP or just needs testing?

**Comment by uyjulian** [comment]
Needs testing, both before video init and after video init..

File: src/render/vitagxm/SDL_render_vita_gxm_tools.c (hunk 1 above)
Comment (uyjulian) [reply src/render/vitagxm/SDL_render_vita_gxm_tools.c]: I'm planning on checking this out

Comment (isage) [reply src/render/vitagxm/SDL_render_vita_gxm_tools.c]: Well, i mean that's only for devunits, to sync perf stats

File: src/video/vita/SDL_vitamessagebox.c [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        {
+            SDL_zero(updateParam);
+
+            updateParam.renderTarget.colorFormat    = dispparam.pixelformat;
+            updateParam.renderTarget.surfaceType    = SCE_GXM_COLOR_SURFACE_LINEAR;
+            updateParam.renderTarget.width          = dispparam.width;
+            updateParam.renderTarget.height         = dispparam.height;
+            updateParam.renderTarget.strideInPixels = dispparam.pitch;
+
+            updateParam.renderTarget.colorSurfaceData = dispparam.base;
```
Comment (isage) [inline src/video/vita/SDL_vitamessagebox.c]: Shouldn't that be `buffer_for_common_dialog[back_buffer_index_for_common_dialog]`?

File: src/render/vitagxm/SDL_render_vita_gxm_tools.c [hunk 3]
```diff
@@ -1178,6 +1178,77 @@ gxm_texture_set_filters(gxm_texture *texture, SceGxmTextureFilter min_filter, Sc
...
+{
+    sceDisplayGetFrameBuf(NULL, SCE_DISPLAY_SETBUF_NEXTFRAME);
+    sceGxmTerminate();
+}
+
+void gxm_init_for_common_dialog(void)
+{
+    for (int i = 0; i < VITA_GXM_BUFFERS; i += 1)
+    {
+        buffer_for_common_dialog[i].displayData.wait_vblank = true;
```
Comment (crudelios) [inline src/render/vitagxm/SDL_render_vita_gxm_tools.c]: This line is erroring for me during build using VitaSDK:

```
SDL_render_vita_gxm_tools.c: In function 'gxm_init_for_common_dialog':
SDL_render_vita_gxm_tools.c:1213:63: error: 'true' undeclared (first use in this function)
 1213 |         buffer_for_common_dialog[i].displayData.wait_vblank = true;
```
Suggested fix:

```suggestion
        buffer_for_common_dialog[i].displayData.wait_vblank = SDL_TRUE;
```

File: src/video/vita/SDL_vitamessagebox.c [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+{
+    SceCommonDialogConfigParam commonDialogConfigParam;
+    SceMsgDialogParam param;
+    SceMsgDialogUserMessageParam msgParam;
+    SceMsgDialogButtonsParam buttonParam;
+    SceDisplayFrameBuf dispparam;
+    
+    SceCommonDialogUpdateParam updateParam;
+    SceMsgDialogResult result;
+    bool setup_minimal_gxm = false;
```
Comment (crudelios) [inline src/video/vita/SDL_vitamessagebox.c]: Another compiler error on VitaSDK:

```suggestion
    SDL_bool setup_minimal_gxm = SDL_FALSE;
```

File: src/video/vita/SDL_vitamessagebox.c [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    }
+    param.userMsgParam = &msgParam;
+
+    dispparam.size = sizeof(dispparam);
+
+    // Setup display if it hasn't been initialized before
+    if (sceDisplayGetFrameBuf(&dispparam, SCE_DISPLAY_SETBUF_NEXTFRAME) < 0)
+    {
+        gxm_minimal_init_for_common_dialog();
+        setup_minimal_gxm = true;
```
Comment (crudelios) [inline src/video/vita/SDL_vitamessagebox.c]: The last compiler error:

```suggestion
        setup_minimal_gxm = SDL_TRUE;
```

**Comment by crudelios** [comment]
@uyjulian I just tested both before and after `video_init` (compiling with DolceSDK this time) and in both cases nothing shows. The app doesn't crash but it is as if the messagebox code is never called.

My code was simple enough:

`SDL_ShowSimpleMessageBox(SDL_MESSAGEBOX_INFORMATION, "test!", "test 1", NULL);`

Also, using DolceSDK compilation fails in `SDL_vitamessagebox.c` because both the struct `SceCommonDialogConfigParam` and the function `sceCommonDialogSetConfigParam` have been removed. Commenting lines `35`, `54` and `55` in that file allowed compilation to finish.

**Comment by uyjulian** [comment]
There are still improvements to be made in this code, and I should have some time this weekend to implement them. I'll also change over to vitasdk from dolcesdk.

**Comment by uyjulian** [comment]
I fixed the remaining issues and made it compile with VitaSDK.  
The following program tests the message box function before and after SDL_Renderer initialization: 
 https://gist.github.com/uyjulian/2d48613c05cf89368efd8e1f61e4bf08

File: src/video/vita/SDL_vitamessagebox.c [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    SceCommonDialogConfigParam commonDialogConfigParam;
+    SceMsgDialogParam param;
+    SceMsgDialogUserMessageParam msgParam;
+    SceMsgDialogButtonsParam buttonParam;
+    SceDisplayFrameBuf dispparam;
+    
+    SceMsgDialogResult dialog_result;
+    SceCommonDialogErrorCode init_result;
+    SDL_bool setup_minimal_gxm = SDL_FALSE;
+#if !SDL_VIDEO_RENDER_VITA_GXM
```
Comment (isage) [inline src/video/vita/SDL_vitamessagebox.c]: I suggest wrapping whole function body in #if/#else instead.

Comment (uyjulian) [reply src/video/vita/SDL_vitamessagebox.c]: Done.

Comment (isage) [reply src/video/vita/SDL_vitamessagebox.c]: Um. Can you switch them around, so there's no `!`? That'll look cleaner. And add unused hint for func params in non-gxm case (` (void)param` etc.)?

Comment (uyjulian) [reply src/video/vita/SDL_vitamessagebox.c]: Sure, I did that

**Comment by isage** [comment]
Thanks, I'll take a closer look/test tomorrow

**Comment by uyjulian** [comment]
A little note on using the message box when the display is initialized.  
The message dialog will render on its own framebuffer with the background initialized to black (instead of the previous frame).  

Rendering the message box on the previous frame would require copying the framebuffer to clear it. This requires copying the  previous frame's framebuffer on the message box framebuffer.  

However, I've found that using `memcpy` to copy the previous frame's framebuffer results in a drop of FPS for the message dialog animation. There may be another way to copy the framebuffer, but I'm not familiar with the GXM API enough, so I won't implement this myself at the moment.  

Other than the above, there shouldn't be any other issues.

**Comment by isage** [comment]
I think that's ok to keep background black.

**Pull request merged**